Main project discussion thread

nilaymaj opened the pull request.

nilaymaj committed:
Add basic global commands API

nilaymaj committed:
Add basic UI component with some commands

nilaymaj committed:
Remove dialog backdrop

NilayMajorwar commented:
@4ian

I've added a basic UI component for the command palette - here's a screenshot (on Nord theme :heart: )

![cmdpalette](https://user-images.githubusercontent.com/18032938/83563615-01caac80-a539-11ea-9dc7-11ac99623154.png)

I've added three commands for now - converting some Mainframe functions into callbacks in the process, and all three are working correctly - properly registering and deregistering (although the Close Project one *might* be updated more frequently than necessary - I'm not sure). 

Added a new option to the debug toolbar dropdown for now, as you suggested. The "Command" box is for the search input (just a dummy for now, though).

Please try it and point out any changes or improvements :) In the meantime, I'll continue with adding more commands, I guess.

nilaymaj committed:
Remove unnecessary import

NilayMajorwar commented:
I've added all the global commands from the Trello list in Mainframe. Everything seems to be working well, but five of the commands' handlers are updated on every Mainframe rerender - I've marked these on Trello (see the "Commands added" list). I'll start with fixing them now.

4ian commented on newIDE/app/src/CommandPalette/CommandHooks.js:
```diff
@@ -0,0 +1,17 @@
+  const manager = React.useContext(CommandsContext);
+  React.useEffect(
+    () => {
+      if (cmdOpts.enabled) {
```
"uber nitpicking": prefer to reduce indentation and early return: `if (!cmdOpts.enabled) return;`

4ian commented on newIDE/app/src/CommandPalette/CommandPalette.js:
```diff
@@ -0,0 +1,68 @@
+
+const CommandPalette = (props: Props) => {
+  const { onClose, open } = props;
+  const mng = React.useContext(CommandsContext);
```
nitpicking: don't use abbreviation in the code, they make it most of the time harder to read ("mng" is it for manager or management) and make it harder to understand (what's a manager").

Instead, name your variables as your type unless you have a good reason not too :) Here the type returned is a "CommandManager", so this variable can be called `commandManager`.

4ian commented on newIDE/app/src/MainFrame/Toolbar/index.js:
```diff
@@ -15,6 +15,7 @@ type Props = {|
   requestUpdate: ?() => void,
   simulateUpdateDownloaded: ?() => void,
   simulateUpdateAvailable: ?() => void,
+  openCmdPalette: () => void,
```
nitpicking: avoid abbreviations: `openCommandPalette`.

4ian commented on newIDE/app/src/MainFrame/Toolbar/index.js:
```diff
@@ -83,6 +84,13 @@ export class MainFrameToolbar extends PureComponent<Props, State> {
+                  disabled: false,
+                  click: () => {
+                    this.props.openCmdPalette();
+                  },
```
nitpicking: if you don't need an extra closure, just do: 

```suggestion
                  click: this.props.openCmdPalette,
```

4ian commented on newIDE/app/src/MainFrame/index.js:
```diff
@@ -221,6 +223,7 @@ const MainFrame = (props: Props) => {
   const preferences = React.useContext(PreferencesContext);
   const [previewLoading, setPreviewLoading] = React.useState<boolean>(false);
   const [previewState, setPreviewState] = React.useState(initialPreviewState);
+  const [cmdPanelOpen, openCmdPanel] = React.useState<boolean>(false);
```
nitpicking: `[commandPanelOpen, openCommandPanel]`

4ian commented on newIDE/app/src/MainFrame/index.js:
```diff
@@ -959,6 +975,23 @@ const MainFrame = (props: Props) => {
     autosaveProjectIfNeeded();
   };
 
+  useGlobalCommand('LAUNCH_PREVIEW', {
```
some thoughts on where to place commands in component code: might be a good idea to place them at single place, before the `return <blabla>`.

In a way, these hooks are "rendering" things. It's not DOM or components, they are rendering commands :) 
I suggest them to be after the logic hooks, so that a component is:

function MyComponent(props) {
  //Logic and hooks like useState()

  // useGlobalCommand hooks (the "commands" that are a bit part of the UI when you think about it)
  // return <Dialog>...</Dialog> (the UI shown on screen)
}

nilaymaj replied:
Yeah, I was thinking about that actually! Will do that, sure.

4ian commented on newIDE/app/src/MainFrame/index.js:
```diff
@@ -1355,21 +1423,34 @@ const MainFrame = (props: Props) => {
+    [props.unsavedChanges, state.currentProject, props.i18n, closeProject]
+  );
+
+  useGlobalCommand('CLOSE_PROJ', {
```
nitpicking: `CLOSE_PROJECT`

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+    console.warn('Initialised command manager');
+  }
+
+  registerGlobal = (cmdName: string, cmdOpts: GlobalCommand) => {
```
Let's give a more precise name. What are we registering? A command :)
`registerCommand`.

4ian replied:
nitpicking again (sorry :D): `commandName` and `command`.

`cmdOpts` is unfortunately not a proper name because it's an abbreviation for "command options"... but this argument is not an "option for commands". It **is** a Command (with a big C, like the type).

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+|};
+
+export default class CommandManager {
+  globalCommands: { [string]: GlobalCommand };
```
Should we really make the distinction between "global" commands and not global?
After all, nothing is really global (even "global variables" should have been called "game variables" ;)). 
Maybe in the future we'll allow multiple instances of the editor to run or multiple projects to be opened, in which case commands could be specified to the project.

I would recommend to go simply with:
`commands: { [string]: Command }`

nilaymaj replied:
I named them `globalCommands` to differentiate them from `gotoCommands` (or `compoundCommands`) which will be added later. Although yes, "global" isn't the right word - "basic" or "direct", maybe?

nilaymaj replied:
I've refactored everything to just `*Command` now (removed the "global"), since in any case adding "global" right now wasn't very relevant as that's the only kind of commands in the code currently. What should I name them, when I start with adding goto commands?

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+// @flow
+type CmdHandler = () => void | Promise<void>;
```
nitpicking: `CommandHandler`

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+  handler: CmdHandler,
+|};
+
+export type UIGlobalCommand = {|
```
What do you think of `NamedCommand`? As it's a command, with a name.

nilaymaj replied:
Ah right, that's definitely better!

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+// @flow
+type CmdHandler = () => void | Promise<void>;
+
+export type GlobalCommand = {|
```
See my other comment below, I think we could name this simply "Command". A command has a display text for the user, can be enabled or not, and has a function to be run. Seems good to me, but it's not global in any sense (it's the usage that you're making of it that determines if it's global or not).

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+    console.warn(`Command ${cmdName} unregistered!`);
+  };
+
+  getGlobalCommands = () => {
```
`getAllCommands`?

4ian replied:
or `getAllNamedCommands` actually

4ian commented on newIDE/app/src/CommandPalette/CommandHooks.js:
```diff
@@ -0,0 +1,17 @@
+import CommandsContext from './CommandsContext';
+
+export const useGlobalCommand = (cmdName: string, cmdOpts: GlobalCommand) => {
+  const manager = React.useContext(CommandsContext);
```
nitpicking: `const commandManager = `

(see other comments for justification).

4ian commented on newIDE/app/src/CommandPalette/CommandPalette.js:
```diff
@@ -0,0 +1,68 @@
+import ChevronRightIcon from '@material-ui/icons/ChevronRight';
+import CommandsContext from '../CommandPalette/CommandsContext';
+
+type Props = {
+  open: boolean,
+  onClose: () => void,
+};
```
Always always use the burger notation in Flow: `{| |}`. Sadly it's not the default in Flow (should be in a future version). Otherwise Flow won't warn if you use put extra props (or make a mistake in a name).
```suggestion
type Props = {|
  open: boolean,
  onClose: () => void,
|};
```

4ian commented on newIDE/app/src/MainFrame/index.js:
```diff
@@ -529,43 +532,56 @@ const MainFrame = (props: Props) => {
-  };
+  }, []);
+
+  useGlobalCommand('OPEN_PROJ_MNG', {
```
`OPEN_PROJECT_MANAGER`

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,43 @@
+type CmdHandler = () => void | Promise<void>;
+
+export type GlobalCommand = {|
+  displayText: string,
```
We'll need this to be a `MessageDescriptor`: `displayText: MessageDescriptor`.

So that we can register commands as translatable, using the `t` macro from js-lingui:
```
displayText: t`Open project manager`
```

and finally use `i18n._(displayText)` to do the translation in the UI

nilaymaj replied:
Oh right sorry, I forgot about that! :sweat_smile: Will add it in a while

4ian commented on newIDE/app/src/CommandPalette/CommandPalette.js:
```diff
@@ -0,0 +1,68 @@
+  return (
+    <Dialog
+      onClose={handleClose}
+      aria-labelledby="command-palette"
```
Great idea to put the aria label here! 👍

nilaymaj replied:
Thanks!

4ian reviewed (commented):
That's great progress! :D 
I've not yet tried it but I read the code. :) As we already discussed the overall approach looks good.

I wrote tons of nitpicking, because I'm very picky (sorry ;)) but also because it's important to get the convention good now to avoid extra work later. 

Overall this is a very good start. Looking forward to what is coming next 👍👍

nilaymaj committed:
Review fixes

nilaymaj committed:
Some more review fixes

NilayMajorwar commented:
@4ian 

I started with fixing the unnecessary handler updates, and turns out it will need a little bit more than just adding React.useCallbacks. For example, some functions use things that are defined *after* that function, and adding those things to dependency lists means I'll have to reorder the functions so that they're used after being declared. Also, every function that is somehow called in any of the command handlers will have to be wrapped in a useCallback to prevent unnecessary handler updates - so maybe around 10-15 (or 20) useCallbacks will be added.

Should I proceed with this? Otherwise, I can focus on finishing the UI part first, then move on to fix these issues.

nilaymaj committed:
Fix unnecessary updates due to t macro

nilaymaj committed:
Add MUI autocomplete to palette UI

NilayMajorwar commented:
![snap](https://user-images.githubusercontent.com/18032938/84150960-3d58ff80-aa80-11ea-831f-52a9525d7a19.png)

nilaymaj committed:
Position palette to top, minor clean up

nilaymaj committed:
Update Mainframe

nilaymaj committed:
Fix warning

nilaymaj committed:
Add temporary kbd shortcut for palette

NilayMajorwar commented:
I've added a very temporary keyboard shortcut for opening the command palette (see previous commit 07727fc2f248443714cc420c4079b41bcf9c78af). It's just a useEffect in Mainframe. The code looks pretty ugly - it's more like "yeah it works" than actual code.

The vanilla JS logic is a bit different than before. I realized that there was an issue with the original approach I mentioned. The problem was that clicking on the empty space in properties panel, etc led to focus somehow going to the body tag, which is outside `div.main-frame`. To fix that, now I also allow opening the palette if the active element is the body tag. Not completely pretty, but seems to work pretty well now.

4ian commented on newIDE/app/src/MainFrame/index.js:
```diff
@@ -1665,6 +1670,98 @@ const MainFrame = (props: Props) => {
+    };
+    document.addEventListener('keyup', handler);
+    return () => document.removeEventListener('keyup', handler);
+  }, []);
```
Isn't there a dependency missing (at least openCommandPalette)?

nilaymaj replied:
Actually no, nothing's missing. No warnings either :)
I was surprised about it too, when omitting `openCommandPalette` did not raise a warning - I guess the ESLint hooks rule has recently started to consider that the "setSomething" functions coming directly from `React.useState` don't really change themselves.

4ian replied:
That's great then 👍

4ian commented:
I think it's fine for now! Would be great to move this to a custom hook provided  `newIDE/app/src/CommandPalette/` (I let you choose if you want to create a new file or use an existing one).

nilaymaj committed:
Move kbd shortcut to custom hook

nilaymaj committed:
Fix hook name

NilayMajorwar commented:
Done. I've moved the code to a custom hook in `.../CommandPalette/CommandHooks.js` for now - the code is specific to the command palette anyway, due to the `Ctrl+P` keyboard shortcut being hardcoded into the hook. 

I'll probably move it to a more appropriate location (and make it more flexible) when I start with shortcuts, since this piece will be reused then to enable and disable shortcuts according to whether an overlay is open.

4ian commented:
> I'll probably move it to a more appropriate location (and make it more flexible) when I start with shortcuts, since this piece will be reused then to enable and disable shortcuts according to whether an overlay is open.

Yes, we'll see how to handle customizable shortcuts later (we'll need them here, in commands, in KeyboardShortcuts.js and a few other places). It's fine to use a hardcoded shortcut for now because everything else is hardcoded.

Shortcuts will probably be stored in preferences, with maybe a context/hook to get the shortcut and use it in a command/KeyboardShortcuts.js (in the most appropriated format).

nilaymaj committed:
Close dialog on single Esc keypress

nilaymaj committed:
Fix delay between options and textbox display, improve UI

nilaymaj committed:
Fix warning

NilayMajorwar commented:
I've made a few changes:
- Made changes to the palette UI so that a single `Esc` keypress now closes the palette.
- The Autocomplete is no longer wrapped in DialogTitle - this tweaks the UI in a pretty interesting way, actually:
(left: old, right: new)
![image](https://user-images.githubusercontent.com/18032938/84795094-adcdc680-b014-11ea-867a-1c778b514c24.png)
- There was an issue that I forgot to mention earlier - the dropdown appeared fraction of a second before the dialog (and input box) appear, and it didn't look very smooth. It was happening because the dialog had a fade transition by default, while the dropdown is rendered by portal and thus appears instantaneously. To fix this, I've removed the dialog fade transition (made transition duration 0s).

Next, I'll try to add a preference flag for the palette. I was thinking of checking for the preference flag inside the keyboard shortcut hook I added, and ignoring shortcut-press if flag is disabled - a drawback is that command handlers will be registered and deregistered in the background even if the command palette is disabled, but I can't find an easy way to disable the whole command system with a preference flag.

nilaymaj committed:
Add preference flag for command palette

nilaymaj committed:
Clean up - remove debug warnings and toolbar item

NilayMajorwar commented:
I've added a preference flag for using the command palette (disabled by default), and removed debug-related things (unnecessary `console.warn`s, and the toolbar debug menu item for opening palette). Checked everything, the commands are working, palette opens only when the flag is enabled and no dialog/overlay is open, and no warnings. I'll open a draft PR for testing.

NilayMajorwar commented:
Oh wait, @4ian there's a bit of a serious issue with the keyboard shortcut - it won't work on browsers, since `Ctrl+P` is an internal browser shortcut for printing the current page!

nilaymaj committed:
Merge branch 'master' of github.com:4ian/GDevelop into cmd-palette

4ian commented:
> it won't work on browsers, since Ctrl+P is an internal browser shortcut for printing the current page!

Do you think there is a chance for us to override this? I don't think so, can we support another shortcut in addition to this one?

>  a drawback is that command handlers will be registered and deregistered in the background even if the command palette is disabled, but I can't find an easy way to disable the whole command system with a preference flag.

I think it's no problem at all, having an event handler for the shortcut even if it's actually doing nothing should not be harmful. 

> this tweaks the UI in a pretty interesting way, actually:

Nice, it looks better indeed!

4ian commented on newIDE/app/src/CommandPalette/CommandPalette.js:
```diff
@@ -0,0 +1,87 @@
+            renderInput={params => (
+              <TextField
+                {...params}
+                placeholder="Start typing a command..."
```
We need to add translation marker to this :)
Is the placeholder a React node or a string?
If a React Node, use `<Trans>` component. Otherwise, call 
```
i18n._(t``)
```

to get a string to pass to the placeholder

nilaymaj replied:
Right, forgot about that one. Done.

nilaymaj committed:
Add i18n marker to input placeholder

NilayMajorwar commented:
> can we support another shortcut in addition to this one

We can, yes, but the choices aren't too many, considering in-built shortcuts from most major browsers. The following seem available on Chrome and Firefox:
```
Ctrl + Shift + F
Ctrl + Shift + L
Ctrl + Shift + U
Ctrl + Shift + B
```

This also highlights the problem we're going to face while adding shortcuts - there *just* aren't enough unused keyboard shortcuts available in a browser to cover all of GDevelop's functions. A lot of sites go for single-letter or double-letter shortcuts (like "<kbd>g</kbd>, then <kbd>p</kbd>") instead, which will require a little effort to avoid clashing with text inputs, etc.

4ian commented:
I've also seen things like "double shortcuts" like `Ctrl+/` then followed by the letter, like `p`

Related to this, `Ctrl+/`, or a single letter shortcut like `?`, can bring on the screen a "shortcut list", showing all shortcuts currently available. This could be a variation of the CommandPalette, with some props set to only show the commands with shortcuts (and show the shortcut in the list). It's good for discovery, as people just have to remember a single shortcut to get a list of all they can use (the command palette/shortcut palette can also be accessible from the main menu in Electron). 

> there just aren't enough unused keyboard shortcuts available in a browser to cover all of GDevelop's functions.

That's ok. We'll prioritize first shortcuts that make sense/are popular in other apps. It's a pity that Ctrl/Cmd+P is not available (there is no way to prevent the print shortcut to work I guess, in the major browsers?).
Is adding "shift" enough to make it work, i.e is Ctrl + Shift + P (or Cmd + Shift + P) working? On my mac on Chrome, it seems to be available (it's actually used by GitHub to toggle the preview while typing this message! :D).
Knowing that Ctrl/Cmd + Shift + P is also used in VSCode to bring the command palette (Ctrl/Cmd + P is the "goto anything"), it makes totally sense to have both of these shortcuts. Only one will work in the web-app, but that's ok.

NilayMajorwar commented:
> shortcut list

That's a great idea - I've added it to the Trello board.

About the shortcuts:

Okay `Ctrl+Shift+P` works on GitHub indeed - it opens a more detailed print dialog when this message input isn't focused, so they have definitely overridden this - probably I can too, then. I did try `e.preventDefault()` for `Ctrl+P` but now I realize I did it for keyup events, while inbuilt functions are bound on keydown. I'll try overriding it again.

nilaymaj committed:
Prevent browser print shortcut

NilayMajorwar commented:
I tried `e.preventDefault`, and for `Ctrl+P`, it works in both Chrome and Firefox. For `Ctrl+Shift+P`, `e.preventDefault` works on Chrome but not on Firefox. Even GitHub's preview shortcut does not work on Firefox.
I've pushed a commit for overriding the browser's print function on `Ctrl+P`.

4ian commented:
Sounds good! Thanks for researching into this. That sounds satisfying then :)

nilaymaj committed:
Auto highlight first command in dropdown

nilaymaj committed:
Move mainframe commands to custom hook

nilaymaj committed:
Add fuzzy search logic

nilaymaj committed:
Fix kbd shortcut hook, and review fix

nilaymaj committed:
Add comments

nilaymaj committed:
Fix and remove eslint-disable-next-line

NilayMajorwar commented:
Sorry for the delay - had to sort out some minor issues to get things up and running here :sweat_smile:.
- I've pushed a commit that moves the t macro strings out of the custom hook (`MainFrameCommands.js`) and into the open, which then fixes the need for eslint-disable-next-line.
- About adding "open command palette" item to the electron app menu: it's slightly tougher than I expected, since I'd need to reuse the "disable when dialog open" logic to dynamically disable the electron menu item too. I'll think about how to do that.

I'm starting out with compound commands now.

4ian commented:
Nice! 
For adding it to the electron app menu, no big deal let's not loose too much time on this. 

Looking forward to see how compound commands will be working! :)

nilaymaj committed:
[wip] Add basic compound command functionality

NilayMajorwar commented:
Pushed a commit. It's not much really, a bit half-baked and ugly right now, there are things to debug too. So don't review it just yet :sweat_smile:. 

I've added the scoped command context provider (for only compound commands), and it's working, but there seem to be some minor issues like too many command manager instantiations - don't know where that's coming from, since it seems to function correctly. I'll look into that.

Right now, my goal is to just implement compound commands for the current scene/tab - so basically the user can use the command palette to open the edit dialog for an object from the current scene only. I initially looked to implement "cross-project" compound commands, but realised after 2-3 days that there are some possibly significant roadblocks with our current implementation idea that need to be discussed. So I instead went for just implementing it for the currently open scene/tab.

4ian commented:
> but there seem to be some minor issues like too many command manager instantiations - don't know where that's coming from, since it seems to function correctly.

I quickly read through the code and found `React.useRef(new CommandManager(true));` => this will create a new CommandManager at very render (because `useRef` is no more and no less than a function taking a parameter, so the parameter here will be instanciated). I think the solution is to use useRef(null) and then create the ref if necessary, see this: https://github.com/facebook/react/issues/14490

> Right now, my goal is to just implement compound commands for the current scene/tab - so basically the user can use the command palette to open the edit dialog for an object from the current scene only.

That's perfectly fine. We'll add commands to "switch tabs" and also commands to open a scene (or anything that you can open from the project manager) so that if you want to edit an object of scene, you can still first launch the command to open the scene/tab containing it, then launch the command to edit the object.

Of course we can add later the "goto" commands that would do this automatically for you, but let's go step by step.
The goal would be to be able to do with the commands most actions that you can do with a mouse.

NilayMajorwar commented:
![editobj](https://user-images.githubusercontent.com/18032938/86236859-63be1680-bbb8-11ea-8617-cf9db87a3b7c.gif)

> useRef is no more and no less than a function taking a parameter, so the parameter here will be instanciated at every render

Yup, that was the thing - I should've noticed that :sweat_smile: 

- The scoped provider is working well, the "Edit Object" command is not shown on non-scene tabs and the options shown are only from the open scene.

- I haven't yet focused on making the command properly work with the ObjectsList render cycle, so there are issues like:
  - The command does not register at all if the objects panel is hidden when a project is opened (even if the objects panel is opened later)
  - Creating new objects does not seem to update the command options until you go to another tab and back.
  - ...among other similar issues

- For now, I've just focused on making it work, so there is quite a bit of code redundancy - for example, the command picker autocomplete and the option picker autocomplete are currently stored in two separate files even though the code is almost same. I'll work on improving those things once the command update issues is solved.

nilaymaj committed:
Add basic compound command functionality

Improve palette UI to support option dropdown.
Add a pretty buggy Edit Object compound command.
Fix too many instantiations of command manager.

nilaymaj committed:
Move Edit Object command to SceneEditor

nilaymaj committed:
Merge command, option pickers into single component

nilaymaj committed:
Shift to dynamic option generation method

NilayMajorwar commented:
Just a mini progress report:
- I've moved the Edit Object command to the SceneEditor component, and it's working nice.
- The code redundancy I mentioned earlier (command picker and option picker) is now fixed -  the flow typing probably needs a review though.
- I tried out the option-generating function method, and you're right, it works perfectly! The project and layout object don't change, so the option generation method ensures that the generated options are *absolutely* correct.

I'll proceed with adding more compound commands now!

nilaymaj committed:
Add Edit Object Variables compound command

4ian commented:
Sounds good! 👍 I'll take a look at the code

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,103 @@
+  displayText: MessageDescriptor,
+  enabled: boolean,
+  generateOptions: () => Array<CompoundCommandOption<T>>,
+|};
```
Two questions:
* Could we simply rename `CompoundCommand` to `CommandWithOptions`. Compound seems a bit complex for someone that do not have the context, and after all it's no more and no less than a command that has a bunch of options :)
* Going a bit further, could we have a single `Command` type, but with an optional `generateOptions`? (`generateOptions?: () => ...`)

Trying to understand if we really need to make the distinction between these.

4ian replied:
Note that in Flow you can also do something like:

```
export type Command = {|
  displayText: MessageDescriptor,
  enabled: boolean,
  handler: CommandHandler,
|} | {|
  displayText: MessageDescriptor,
  enabled: boolean,
  generateOptions: () => Array<CompoundCommandOption<T>>,
|}
```

Though this can somewhat leads to complex error messages. But basically this tells Flow that a command is either the first shape or the other, so it can have handler OR a generateOptions but not both - and it will force you to check this when needed.

nilaymaj replied:
Okay, that's what I was about to suggest, in place of an optional `generateOptions` :D 
I'll do this right away!

nilaymaj replied:
Did this, and removed the redundancy from the command manager functions. I couldn't merge the hooks into one, though. I tried something like this:

```js
// Attempted unified hook:
const useCommand = (commandName: string, command: Command) => {
  const commandManager = React.useContext(CommandsContext);
  const { displayText, enabled } = command;

  React.useEffect(() => {
    if (!enabled) return;
    if (command.handler) {
      // Direct command
      commandManager.registerCommand(commandName, {
        displayText, enabled, handler: command.handler
      })
    } else {
      // Command with options
      commandManager.registerCommand(commandName, {
        displayText, enabled, generateOptions: command.generateOptions
      })
    }
  }, [commandManager, commandName, displayText, enabled, command.handler, command.generateOptions])
}
```

The issue is that we get stuck between React warnings and Flow errors in the above code. Flow won't allow us to add `command.handler` or `command.generateOptions` to the dependency list because each of those properties are absent in one or the other of the enum types of `Command`, while omitting any of those from the dependency list or using React hooks in not the top-level of custom hook is not right.

4ian commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -0,0 +1,103 @@
+    delete this.commands[commandName];
+  };
+
+  registerCompoundCommand = <T>(
```
In the same way, I think we should see if we can have the commands registered in the same way, be it a command or a command with options.
At least, commands are indexed by names, and I see no reason for allowing a command and a command with options that have the same name to be allowed. Unfortunately, currently as we're storing them in separate objects, this situation can arise.

Let's see if we can store them in a single object?

nilaymaj replied:
Done.

4ian commented on newIDE/app/src/SceneEditor/index.js:
```diff
@@ -1028,6 +1029,43 @@ export default class SceneEditor extends React.Component<Props, State> {
+            displayText: t`Edit object...`,
+            enabled: true,
+            generateOptions: () =>
+              enumerateObjects(project, layout).containerObjectsList.map(
```
Looks good 👍

4ian commented on newIDE/app/src/SceneEditor/index.js:
```diff
@@ -1028,6 +1029,43 @@ export default class SceneEditor extends React.Component<Props, State> {
+            displayText: t`Edit object variables...`,
+            enabled: true,
+            generateOptions: () =>
+              enumerateObjects(project, layout).containerObjectsList.map(
```
if we move this to a hook/component "UseSceneEditorCommands", we can probably factor this function in something called `generateLayoutObjectsOption({project, layout, onChoose: (object) => this.editObjectVariables(object)})`.

(just to avoid this tiny bit of repetition, though this would be really worth factoring if there are more than 2 usages of it).

nilaymaj replied:
Done. I've moved the commands to a separate file like in Mainframe, seems cleaner.

nilaymaj committed:
Merge compound command flow type into command

nilaymaj committed:
Move scene editor commands to separate file, plus minor changes

nilaymaj committed:
Add Open Recent Project command, and minor fixes

nilaymaj committed:
Add project manager related commands

NilayMajorwar commented:
I've added the following commands to the Project Manager:
- Open scene...
- Open external layout...
- Open external events...
- Open extension...

I've also added the "Open recent project..." command to ElectronMainMenu. It updates on every Mainframe rerender though, since the Mainframe function used for opening recent project (`openFromFileMetadataWithStorageProvider`) is not wrapped in a useCallback. Wrapping this one is simple though, since it does not seem to have any non-useCallback dependency functions. Should I push a commit here itself for wrapping this or make a separate PR?

Similar to Mainframe and Scene Editor, I've added the project manager commands in a separate component `UseProjectManagerCommands`  and used a helper function to reduce redundancy in `generateOptions`. Everything seems to work fine mostly - there's just minor issues (I can't seem to consistently reproduce it though):
- Sometimes there's a scene tab in the editor that does not seem to belong to the project - opening the tab shows `No layout named "Game" found`.
- On opening a scene via the command, sometimes it opens new tabs for the scene while sometimes it closes the current scene and opens the new scene. Is this expected?

Both of the above happen randomly - I'll try to reproduce them consistently and see what causes it.

Also, except the above issues and the goto anything part, all the commands from the command list on Trello are now done!

4ian commented:
>  Should I push a commit here itself for wrapping this or make a separate PR?

I think it's fine to do it in this PR, the goal will be to merge this PR sooner than later :)

> Sometimes there's a scene tab in the editor that does not seem to belong to the project - opening the tab shows No layout named "Game" found.

Very strange. What's the name of the tab? How do you open it, using a command?
Is it possible that there is a mismatch in the enumeration of layouts/external layout/events maybe?

> sometimes it opens new tabs for the scene while sometimes it closes the current scene and opens the new scene. Is this expected?

I don't expect the current scene to be closed. Would be good to inspect more what is happening (add a few console.log to try to see the state of the tabs and functions called) - there is something fishy there :)

NilayMajorwar commented:
> I think it's fine to do it in this PR, the goal will be to merge this PR sooner than later :)

Okay, I'll push a commit for this, then.

> Very strange. What's the name of the tab? How do you open it, using a command?
Is it possible that there is a mismatch in the enumeration of layouts/external layout/events maybe?

I don't remember clearly. It was certainly a scene editor tab though, named "Game". I think I was trying out the commands I added, but I'm not sure :sweat_smile:. A mismatch in enumerate and onOpen functions seems probable, in which case it might be inadvertently fixed already - it just happened once, after all. I'll still try to make it happen again.

> I don't expect the current scene to be closed. Would be good to inspect more what is happening (add a few console.log to try to see the state of the tabs and functions called) - there is something fishy there :)

Okay then, I'll see what's causing this. It's happening consistently, actually - in fact, there's more: if tabs for scene A and an extension are open, and I use a command to open scene B from the scene A tab, scene B tabs are properly opened but the extension tab closes! This probably extends to external events and layouts too.

nilaymaj committed:
Wrap recent project open function in useCallback

NilayMajorwar commented:
Okay, so the second issue is because the function `onOpenLayout` passed to Project Manager gets outdated (uses an older `editorTabs` object, to be precise) the moment I open another tab. This is because ProjectManager does not rerender if it's hidden, and thus the onOpen handler provided in the command is outdated. All of this seems to arise from the assumption that new tabs are opened only when the Project Manager is open, which is no longer true with the command palette.

Since the openLayout, openExternalEvents etc functions are from Mainframe anyway, I guess I should move the "open [...]" commands to Mainframe instead.

nilaymaj committed:
Move project manager commands to Mainframe

NilayMajorwar commented:
I've moved the project manager commands into Mainframe commands. Everything's working fine now. I still can't reproduce the stale scene editor tab issue (I've tried changing project and everything). I think it was probably due to a mismatch or an outdated `editorTabs` object, in which case it's probably fixed now.

nilaymaj committed:
Add open scene properties command

nilaymaj committed:
Add open scene variables command

nilaymaj committed:
Add edit layer effects command

nilaymaj committed:
Add edit object group command

NilayMajorwar commented:
@4ian I just realised there's an issue with the current implementation. This issue is visible in the "Edit layer effects" command I added: 

If the layers panel is hidden while a project is opened, the "Edit layer effects" command isn't registered - undesirable but expected, since the command hook is used inside the LayersList component which is not rendered. But the command doesn't get registered even after the LayersList is opened by the user.

This is because the scoped provider syncs commands with the central manager only when the tab's `active` status changes. Commands dynamically enabled and disabled on a single tab do not get forwarded to the central manager until the user goes to another tab and back. The scoped provider needs to actively forward and withdraw commands while the tab is open.

I see two possible solutions right now:
- Allow external code to hook into the command manager lifecycle via passing functions like `onCommandRegistered` to CommandManager, that the manager will call after registering/deregistering a command.
- Allow passing an object called `tabCommands` to CommandManager, that represents the scoped manager's commands. Then the central manager will have a direct reference to the scoped provider's commands and can include it when listing registered commands. I'll add functions in CommandManager for forwarding and withdrawing this `tabCommands` object.

The second option seems better to me, mostly since it simplifies things (while the first one adds more things) for fixing this issue. Also, because it's a tiny step *towards* goto anything commands (if it's ever done in the future), since then the scoped provider can somewhat tell the central manager "here are the commands for this tab, and use this function to open the tab if needed". The scoped provider is much closer in code to a tab-opening function than the command manager, so it seems plausible to me (unless I missed something :sweat_smile: ). 

What do you think?

4ian commented:
> Allow external code to hook into the command manager lifecycle via passing functions like onCommandRegistered to CommandManager, that the manager will call after registering/deregistering a command.

Sounds hacky to me, let's avoid this.

> Allow passing an object called tabCommands to CommandManager, that represents the scoped manager's commands.

That's a potential solution, but I'm thinking of a third one ;)

Currently, CommandsContextScopedProvider is exposing the command manager through `scopedManager`, which is a `CommandManager`. It then forward the commands to the upper scope (i.e to what you called `centralManager`) inside the `React.useEffect`, which is triggered when the `active` property changes (note that `scopedManager.commands` will probably never change, because it's an object created once in the CommandManager).

When thinking about the `CommandsContextScopedProvider`, I was thinking of something a bit different. Instead of "forwarding" commands up to the `centralManager` only in some circumstances, you always do it by exposing a "facade", i.e: an object that acts like CommandManager but is implemented by calling into centralManager. Something likes:

```
// In the provider:
const scopedManager = new ScopedCommandManager(centralManager);

// The class:
class ScopedCommandManager {
  constructor(centralCommandManager) { this._centralManager = centralCommandManager; }
  registerCommand(commandName, command) {
    this._centralManager.registerCommand(commandName, command);
  }
  deregisterCommand(commandName) {
    this._centralManager. deregisterCommand(commandName);
  }
  // ... etc for the rest of the interface exposed by CommandManager
}
```

Of course what I did here is for now a bit useless. What is now interesting it that the scopedManager can also retain its own commands, so that its commands are only registered when it's active:


```
class ScopedCommandManager {
  commands: { [string]: Command };
  constructor(centralCommandManager) { this._centralManager = centralCommandManager; }
  registerCommand(commandName, command) {
    this.commands[commandName] = command;
    this._centralManager.registerCommand(commandName, command);
  }
  deregisterCommand(commandName) {
    delete this.commands[commandName];
    this._centralManager. deregisterCommand(commandName);
  }
  // ... etc for the rest of the interface exposed by CommandManager
}
```

So now we've duplicated the commands. The last step is now to integrate a `isActive` boolean and a few functions to register/unregister all commands:

```
// In the provider:
React.useEffect(() => {
  if (props.active) scopedManager.registerAllCommandsInTheCentralManager();
  else scopedManager.deregisterAllCommandsInTheCentralManager();
}, [props.active]);
```

(and maybe you must update `registerCommand` to call the centralManager only if isActive is true, but I can't be sure).

So that you have both:
- transparent registration/unregistration in "real time" of commands.
- unregistration and then registration again of all commands that are in the ScopedCommandManager when the tab goes unactive or active again :)

Does that make sense?

NilayMajorwar commented:
Okay, seems clear. But this would make the Flow types a bit more complicated. Since `CommandManager` and `ScopedCommandManager` will be different types, the commands context will have an enum type (`CommandManager | ScopedCommandManager`). Then although the hooks won't have any issues since `registerCommand` and `deregisterCommand` are available in both manager types, the palette will have to check if the manager obtained from the context is central or scoped (even though we know it will be the central one). 
Is that fine?

4ian commented:
I think we should hide behind an interface: the hook uses a `CommandManager` interface for their props/arguments: https://flow.org/en/docs/types/interfaces/

Then the class `CommandManager` can be renamed to `CentralCommandManager` (just to avoid confusing it with the interface). 
OR maybe better, you name the interface `CommandManagerInterface` and keep `CommandManager` as the class name :)

NilayMajorwar commented:
Oh okay, I didn't know Flow had interfaces too :grinning:. I'll implement this now.

4ian commented:
I usually rarely use classes (apart from React classes) because a lot of things can be done without `Class` in JavaScript thanks to the language being very oriented around functions - actually you could turn your class into a function creating an object with "private" stuff taking advantage of the closures:
```
class MyClass {
  myStuff = {};
  doStuff() { myStuff.test = 123 };
}
```

into: 
```
function createStuff() {
  const myStuff = {};
  return {
    doStuff: () => { myStuff.test = 123 };
  };
}
```

in which case you can just create a single type: `type Stuff = {| doStuff: () => void |}` and `interface` is not needed. I think that's the reason why you don't see it much in Flow typed codebases.
This being said I don't care much, as long as the result is clear, it's fine to use classes like you did and an interface :)

nilaymaj committed:
[wip] Change scoped provider implementation

nilaymaj commented on newIDE/app/src/CommandPalette/CommandsScopedContext.js:
```diff
@@ -9,31 +13,72 @@ type Props = {
-  const scopedManager = useRefInit(() => new CommandManager(true));
+  const scopedManager = useRefInit(
+    () => new ScopedCommandManager(centralManager)
+  );
```
This part gives a Flow error. Since CommandsContext is now typed as containing a `CommandsManagerInterface`, Flow does not know if `centralManager` here is a `CommandManager` or a `ScopedCommandManager`, while the constructor of `ScopedCommandManager` strictly takes a `CommandManager`.

Should I change the type of `ScopedCommandManager._centralManager` as a `CommandManagerInterface`? It kinda hints towards layered facades (like more tabs inside a tab), although that's probably not very useful.

4ian replied:
Yes I think the type should be a `CommandManagerInterface`. The "beauty" of this is that as you mentioned, you can layer the facades - so that you could use a `CommandsContextScopedProvider` inside a `CommandsContextScopedProvider` and it would work perfectly fine.

Not very useful for now, but possible in theory and might get some usage in the future!

nilaymaj commented on newIDE/app/src/CommandPalette/CommandManager.js:
```diff
@@ -34,13 +34,19 @@ export type NamedCommandWithOptions<T> = {|
+
+  registerCommand: (commandName: string, command: Command) => void;
+  deregisterCommand: (commandName: string) => void;
+  getAllNamedCommands: () => Array<NamedCommand>;
```
I added `getAllNamedCommands` to the interface to avoid a Flow error in the palette UI code, which tries to use `commandManager.getAllNamedCommands()` - which would work because we know its going to be the central manager, but Flow doesn't know that. 

Another option would be to use something like `commandManager.getAllNamedCommands ? commandManager.getAllNamedCommands() : []` in the palette UI. Should I go for that instead?

4ian replied:
I would rather avoid the ternary and have it part of the interface, especially if it's possible to offer an implementation for `ScopedCommandManager`.

nilaymaj reviewed (commented):
I've pushed a commit with a tentative version of the suggested implementation for the scoped command provider. It's working properly, but using a common interface is leading to a bit of Flow trouble - please review :sweat_smile:

NilayMajorwar commented:
By the way, now that the scoped provider is almost fixed, the "Edit layer effects" command is only available when the layers panel is open, since the command is added in the LayersPanel component. This was necessary because the Edit Effects dialog state and function are in the LayersList component, not SceneEditor. Should I raise it up to SceneEditor (like the state and function for layer remove dialog)?

4ian commented:
> This was necessary because the Edit Effects dialog state and function are in the LayersList component, not SceneEditor. Should I raise it up to SceneEditor (like the state and function for layer remove dialog)?

I think it would be good to do that, if possible without moving too much or at all the logic/functions into SceneEditor to avoid bloating it. 

In a way, I think of the commands as a secondary interface, that is not shown to the user but "living" in memory. The interface of commands for the SceneEditor is represented by `UseSceneEditorCommands`. This means that we should include here all the commands for the "sub editors" like the objects list (you've done it), the groups list (you've done it), the layers list.

There is still a question of organization: this file will start to contains a lot of commands! Also, wouldn't it be great if we could have commands organized by editors, like components? After all, we have `SceneEditor` component, that contains `ObjectsList`, `ObjectGroupsList`, `LayersList` components, which is very clear and separate the responsibilities... Well we could do the same for the command hooks! :) 
We can have `UseSceneEditorCommands`, that internally uses ("uses" ;)) `useObjectsListCommands`, `useObjectGroupsListCommands`, `useLayersListCommands`.
And these custom hooks can live in a file that is next to the file of the editor.

The advantage of splitting these into custom hooks is that when you're a developer that want to modify LayersList (or any editor) to add a new feature (or modify an existing one), you can directly find the commands of this editor in a file called `LayersListCommands.js` - very clear and colocated at the same place. 

Let me know what you think. Note that here because we want the layers commands (and the other editor commands) to be always available, we call the hook in the SceneEditor. If we want one day to only show these commands when the editor is opened, we can always move the `useLayersListCommands` into the component (and/or we can call it from another parent editor if one day the layers list is shown somewhere else!)

nilaymaj committed:
Change scoped provider implementation

nilaymaj committed:
Move edit layer effects to scene editor commands

nilaymaj committed:
Split scene editor commands into multiple files

NilayMajorwar commented:
Done. Moved the edit layer effects logic into scene editor. Split the scene editor commands into panel-specific parts as you said, and its certainly cleaner now!

NilayMajorwar commented:
The next thing (and maybe last thing related to commands?) is to add a command to ToolbarIcon. Here's what I'm thinking:
- Every command needs a unique identifier, which is tough to get from the existing props of ToolbarIcon (i.e. icon src, tooltip string, accelerator string, disabled, onClick, onContextMenu).
- Some toolbar buttons' tooltips are a bit odd when shown in command palette as the command's display text (the preview button's tooltip has "Launch preview, right click for more", for example). Not sure what to do about that.
- Dropdown toolbar icons are a bit tricky - ToolbarIcon does not have access to the options so can't register a command with options on its own, and adding `useCommandWithOptions` to ElementWithMenu does not seem right to me (or is it?).

4ian commented:
Right there are some limitations, so either we:
- refactor ToolbarIcon to make it "smarter" by adding stuff like a unique identifier, a list of option so that it's then using internally ElementWithMenu (or introduce an intermediate component doing all of that). And probably an optional prop to disable command generation.
- or we decide this is not worth the added complexity and simply add command hooks in the various `Toolbar` components.

While I like the idea of adding a button = automatically adding a command, which is nice on the paper, it's maybe not worth the complexity (especially for options => this would mean adding a new abstraction about menu/options... not a big fan). So let's go for just adding commands in the Toolbar components and call it a day :) It should still result in something that is very easy to maintain and understand, so it's actually a better solution when thinking about it.

nilaymaj committed:
Add toolbar commands for scene editor

nilaymaj committed:
Add toolbar commands for events editor

NilayMajorwar commented:
I've added commands for almost every toolbar button in scene editor and events editor, and it's working fine. For each toolbar, I've created another file `ToolbarCommands.js` in the same folder, that contains all the commands. There's quite a bit of redundancy though, in props and display text - should I move the commands into their respective Toolbar components instead?

Also, I realized a few commands are *still* left - the ones corresponding to items under "Game settings" in Project Manager (Game properties, global variables, etc). Adding them to ProjectManager won't work directly because it gets rerendered only when its opened so dialog-related state updates won't trigger opening the dialog. What should I do about that?

nilaymaj committed:
Fix unnecessary updates for edit layer effects command

nilaymaj committed:
Fix unnecessary updates for some event sheet commands

NilayMajorwar commented:
@4ian I've fixed unnecessary updates for some commands, specifically the "Edit layer effects" command and three event sheet toolbar commands. With that, I don't think there are any more significant "update leaks" for any command. 

Some commands haven't been added, see above message for that.

If there's nothing else left, should I clean up and create a PR on the main repository?

4ian commented:
I think it's a good time for a PR! :) 

> specifically the "Edit layer effects" command and three event sheet toolbar commands. With that, I don't think there are any more significant "update leaks" for any command.

Sounds good, thanks for that!

>  For each toolbar, I've created another file ToolbarCommands.js in the same folder, that contains all the commands. There's a bit of redundancy in props and display text - although moving commands into the toolbar components themselves looks uglier to me.

I'm fine with both :)

> I can add return true to shouldComponentUpdate if any dialog state variable changes. Is that solution okay?

I think that yes it's a good solution. Basically this means that the existing `shouldComponentUpdate` was incomplete in its implementation (checking the state is something that I did in some other components that have a `shouldComponentUpdate`).

> edit: Also, are there any other commands that are to be added now?

If we have commands for all toolbars, for all stuff in the ProjectManager, I think we're in good shape! Let's open a PR and see if we miss commands and add them on the go :)

nilaymaj committed:
Add project manager commands

NilayMajorwar commented:
> well, turns out I used project variable already sweat_smile Fine to keep it then, but let's keep "global variables" in the user facing sentence

Didn't really understand - is the term "project variables" already used for something else in GDevelop?

4ian commented:
> is the term "project variables" already used for something else in GDevelop?

No, it's a proper term for what you've done. It's just that I chose "global variable" for the interface shown to the user and for the documentation ([like here](http://wiki.compilgames.net/doku.php/gdevelop5/tutorials/basic-game-making-concepts)).
Project variable is a synonym for "global variables", and as it's already used in the codebase, it's fine to keep it.

nilaymaj committed:
Fix some command names and display texts

nilaymaj committed:
Merge branch 'master' of github.com:4ian/GDevelop into cmd-palette

nilaymaj committed:
Add grid-related commands

nilaymaj committed:
Review fixes

nilaymaj committed:
Remove unnecessary argument

nilaymaj committed:
Use any instead of *

nilaymaj committed:
Rename generic parameter

nilaymaj committed:
Show command text in option picker placeholder

nilaymaj committed:
Add stories for command palette

nilaymaj committed:
Review fix for palette stories

nilaymaj committed:
Move enabled property from command to hook argument

nilaymaj committed:
Remove unused value property from command option

NilayMajorwar commented:
Closing this one. Opened #3 for work on shortcuts.
